[SPARK-31432][DEPLOY] Make SPARK_JARS_DIR configurable

marblejenka opened the pull request:
### What changes were proposed in this pull request?

This PR makes SPARK_JARS_DIR configurable in scripts under bin/sbin and launcher.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
As I wrote in SPARK-31432, the change is benefitical for flexibility of deployment.

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
No, as long as the user dose not set SPARK_JARS_DIR in the environemnt variable.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
scripts under bin/sbin
- I built release archive by ./dev/make-distribution.sh and run some tests by hand on linux and windows. The log of the test is https://gist.github.com/marblejenka/28cb3e14056e72c8d6e540c999516459

launcher
- I have added a testcase assuming SPARK_JARS_DIR is not set
- It is possible to write a testsase for SPARK_JARS_DIR is set, but it is difficult to do it with portability. I will add the testcase if reviewer requests.

Commit:
allow to customize jars dir

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

maropu commented:
ok to test

maropu commented:
cc: @dongjoon-hyun @srowen

maropu commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -328,4 +329,9 @@ static String findJarsDir(String sparkHome, String scalaVersion, boolean failIfN
+    String jarsDir = System.getenv(ENV_SPARK_JARS_DIR);
+    return jarsDir != null ? jarsDir : join(File.separator, sparkHome, "jars");
+  }
+
```
nit: remove the blank.

marblejenka replied:
This is because most of the code in launcher put a blank line after the last method. Shoud I fix it?

srowen replied:
It's trivial either way. I personally prefer a blank line here as other methods are followed with a blank line.

maropu commented on launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:
```diff
@@ -22,6 +22,7 @@
 import java.util.List;
 
 import org.junit.Test;
+
```
nit: unnecessary change?

marblejenka replied:
Yes, I overlooked it. I will remove it.

SparkQA commented:
**[Test build #121197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121197/testReport)** for PR 28200 at commit [`752e6d4`](https://github.com/apache/spark/commit/752e6d403aa93b7c7345e1466dff0af47f9420e7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25885/
Test PASSed.

planga82 commented:
I think it is very useful, I'm going to open a pull request to introduce it in the spark configuration documentation.
Thanks!

SparkQA commented:
**[Test build #121197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121197/testReport)** for PR 28200 at commit [`752e6d4`](https://github.com/apache/spark/commit/752e6d403aa93b7c7345e1466dff0af47f9420e7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121197/
Test PASSed.

planga82 committed:
Update configuration page

marblejenka committed:
removed unnecessary blank line

SparkQA commented:
**[Test build #121213 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121213/testReport)** for PR 28200 at commit [`a128f01`](https://github.com/apache/spark/commit/a128f0100a23a7a9ad7f6133b1c7704c824270ba).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25901/
Test PASSed.

SparkQA commented:
**[Test build #121213 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121213/testReport)** for PR 28200 at commit [`a128f01`](https://github.com/apache/spark/commit/a128f0100a23a7a9ad7f6133b1c7704c824270ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121213/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #121241 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121241/testReport)** for PR 28200 at commit [`a128f01`](https://github.com/apache/spark/commit/a128f0100a23a7a9ad7f6133b1c7704c824270ba).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25928/
Test PASSed.

SparkQA commented:
**[Test build #121242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121242/testReport)** for PR 28200 at commit [`c2f4972`](https://github.com/apache/spark/commit/c2f4972f2268d96ef5e9825dfd637712c61ddda1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25929/
Test PASSed.

SparkQA commented:
**[Test build #121242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121242/testReport)** for PR 28200 at commit [`c2f4972`](https://github.com/apache/spark/commit/c2f4972f2268d96ef5e9825dfd637712c61ddda1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121242/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #121249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121249/testReport)** for PR 28200 at commit [`c2f4972`](https://github.com/apache/spark/commit/c2f4972f2268d96ef5e9825dfd637712c61ddda1).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25936/
Test PASSed.

SparkQA commented:
**[Test build #121241 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121241/testReport)** for PR 28200 at commit [`a128f01`](https://github.com/apache/spark/commit/a128f0100a23a7a9ad7f6133b1c7704c824270ba).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121241/
Test PASSed.

dongjoon-hyun commented:
Thank you for pinging me, @maropu .

dongjoon-hyun commented on docs/configuration.md:
```diff
@@ -2939,6 +2939,10 @@ To specify a different configuration directory other than the default "SPARK_HOM
 
+# Overriding classpath default directory
+
+To specify a different classpath directory other than the default "SPARK_HOME/jars"  you can set SPARK_JARS_DIR. Any of the services under "SPARK_HOME/sbin" folder will use it.
```
nit, `"  you` -> `",  you`

dongjoon-hyun replied:
Why do we limit to only `sbin` here?

dongjoon-hyun replied:
Like the PR title have `bin` and `sbin`, this PR changes the standalone ExternalShuffleService like the following. We had better mention that.
```
bin/spark-class org.apache.spark.deploy.ExternalShuffleService
```

dongjoon-hyun commented on launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:
```diff
@@ -99,6 +99,11 @@ public void testJavaMajorVersion() {
 
+  @Test
+  public void testGetJarsDir() {
+    assertEquals("/spark/home/jars", getJarsDir("/spark/home"));
```
Hi, @marblejenka .
This seems to add a half of the test coverage.
Since this PR's key contribution is to introduce `ENV_SPARK_JARS_DIR`, shall we have a test coverage for `ENV_SPARK_JARS_DIR`, too?

dongjoon-hyun commented:
Thank you for your first contribution, @marblejenka !
I left a few minor comments.

dongjoon-hyun commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -315,7 +316,7 @@ static int javaMajorVersion(String javaVersion) {
+    File libdir = new File(getJarsDir(sparkHome));
     if (!libdir.isDirectory()) {
       libdir = new File(sparkHome, String.format("assembly/target/scala-%s/jars", scalaVersion));
       if (!libdir.isDirectory()) {
```
Ur, this PR seems to cause a misleading error message when the user give a wrong `SPARK_JARS_DIR`.

In case of a wrong `SPARK_JARS_DIR`, the error message should give the user-given `SPARK_JARS_DIR` instead of `assembly/target/scala-%s/jars`. Could you check this once more, @marblejenka ?

Also, it would be great if we can have a negative test coverage for the above wrong `SPARK_JARS_DIR`.

dongjoon-hyun reviewed (changes requested):
Although the existing error message scheme seems to show the last tried `libdir`,  this PR seems to add more complexity with `SPARK_JARS_DIR` in terms of the error message.

In case of `findJarsDir()`,
1. Old lookup behavior: `sparkHome/jars` -> `assembly/target/scala-%s/jars`
2. New lookup behavior in this PR: `SPARK_JARS_DIR` -> `assembly/target/scala-%s/jars`

Currently, this PR doesn't complain about the wrong `SPARK_JARS_DIR`. I believe the following is more compatible with the existing behavior (and the existing installation).

3. `SPARK_JARS_DIR` -> `sparkHome/jars` -> `assembly/target/scala-%s/jars`

In addition to (3), if we can identify the user's wrong `SPARK_JARS_DIR`, it would be great if we can give `WARN` message.

SparkQA commented:
**[Test build #121249 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121249/testReport)** for PR 28200 at commit [`c2f4972`](https://github.com/apache/spark/commit/c2f4972f2268d96ef5e9825dfd637712c61ddda1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121249/
Test FAILed.

marblejenka commented:
@dongjoon-hyun Thanks for the analysis. Based on your comment, I will work on the following.
- add a test in CommandBuilderUtilsSuite for coverage
- modify the behavior of CommandBuilderUtils#findJarsDir in case SPARK_JARS_DIR is specified but seems not to appropriate

Commit:
add test in CommandBuilderUtilsSuite for coverage
modifiy the error message in CommandBuilderUtils#findJarsDir
modify the behaveor in case SPARK_JARS_DIR is specified but seems not collect

Commit:
modified the documentation

marblejenka commented:
@dongjoon-hyun I pushed the modified code. Note that I have added a dependency for  com.github.stefanbirkner:system-rules https://github.com/stefanbirkner/system-rules to the testing of the environment variable. Since it is licensed Common Public License Version 1.0, I think it is not a problem for spark project.
I also modified the document based on planga82's work. Thank you @planga82 !

SparkQA commented:
**[Test build #121321 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121321/testReport)** for PR 28200 at commit [`e931dd9`](https://github.com/apache/spark/commit/e931dd97c29dc9fecb6d550fbc8a37f492919ee3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26005/
Test PASSed.

SparkQA commented:
**[Test build #121321 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121321/testReport)** for PR 28200 at commit [`e931dd9`](https://github.com/apache/spark/commit/e931dd97c29dc9fecb6d550fbc8a37f492919ee3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121321/
Test PASSed.

dongjoon-hyun commented on pom.xml:
```diff
@@ -928,6 +928,18 @@
         <scope>test</scope>
       </dependency>
+      <dependency>
+        <groupId>com.github.stefanbirkner</groupId>
```
Ur, @marblejenka . If possible, could you avoid to introduce new test dependency? Is it inevitable?

marblejenka replied:
It is not impossible, but we have to do the same thing to test code that uses environment variables. Consequently, we will include a copy of that dependency (specifically https://github.com/stefanbirkner/system-rules/blob/master/src/main/java/org/junit/contrib/java/lang/system/EnvironmentVariables.java , which enables to write environment variable via reflection api) in our test code. 
The other way is giving up coverage, but I think it not desirable.

In my opinion, system-rules is beneficial for testing of environment variables in other usecase. Additionally, there should be no problem with the license, so I would like you to merge this.

dongjoon-hyun replied:
`environmentVariables.set` is used only twice. I guessed the following will be enough for this PR. What do you think?
```java
    try {
      System.setProperty("SPARK_JARS_DIR", "...");
      // test code
    } finally {
      System.clearProperty("SPARK_JARS_DIR");
    }
```

marblejenka replied:
System#getProperties and System#getenv are slightly different. Since I use the environment variable, we can't use the suggested way(I choose System#getenv because SPARK_CONF_DIR also uses environment variable). 
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getProperties--
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getenv--

If adding new dependencies in this PR is unacceptable, I will open another PR aimed at improving test coverage using environment variables.
It seems that some code that uses environment variables is not branch-covered according to a coverage report I created. While improving branch coverage of existing stable code may not be very useful, I don't think it is harmful.

dongjoon-hyun commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -314,18 +319,31 @@ static int javaMajorVersion(String javaVersion) {
     return libdir.getAbsolutePath();
   }
 
+  private static String handleJarsDir(File theDir, boolean failIfNotFound, String message) { 
```
`theDir` -> `dir`.

marblejenka replied:
I will fix it. Thank you.

dongjoon-hyun commented on launcher/src/main/java/org/apache/spark/launcher/CommandBuilderUtils.java:
```diff
@@ -314,18 +319,31 @@ static int javaMajorVersion(String javaVersion) {
+    String envSparkJarsDir = System.getenv(ENV_SPARK_JARS_DIR);
+    if (envSparkJarsDir != null && !envSparkJarsDir.isEmpty()) {
+      File envJarsDir = new File(envSparkJarsDir);
+      return handleJarsDir(envJarsDir, failIfNotFound,
```
`return`? What I suggested is to warn and continue the old lookup behavior.
```
SPARK_JARS_DIR -> sparkHome/jars -> assembly/target/scala-%s/jars
```

marblejenka replied:
I will fix it. Thank you.

dongjoon-hyun commented on launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:
```diff
@@ -99,6 +111,56 @@ public void testJavaMajorVersion() {
+
+    // if we have SPARK_JARS_DIR, use it as jars dir
+    File spefifiedJarsDir = tempFolder.newFolder("jars-dir");
+    environmentVariables.set("SPARK_JARS_DIR", spefifiedJarsDir.getAbsolutePath());
```
`"SPARK_JARS_DIR"` -> `ENV_SPARK_JARS_DIR`.

marblejenka replied:
I will fix it. Thank you.

dongjoon-hyun commented on launcher/src/test/java/org/apache/spark/launcher/CommandBuilderUtilsSuite.java:
```diff
@@ -99,6 +111,56 @@ public void testJavaMajorVersion() {
+      findJarsDir(sparkHome, "2.13", true));
+
+    // if SPARK_JARS_DIR is specified but not exists, throws IllegalStateException
+    environmentVariables.set("SPARK_JARS_DIR", join(File.separator, tmpRoot, "jars-dir-not-exists"));
```
`"SPARK_JARS_DIR"` -> `ENV_SPARK_JARS_DIR`.

marblejenka replied:
I will fix it. Thank you.

dongjoon-hyun commented:
Gentle ping, @marblejenka ~

marblejenka commented:
I will correct the points you have pointed out.
Sorry for the late response.

tgravescs commented:
I'm not sure I understand the description of the use case here, you want to point to a separate JAR directory but keep the same SPARK_CONF_DIR?  I'm trying to understand why not set SPARK_HOME to a different location.

note that on yarn there are other options for jars in HDFS like spark.yarn.jars and spark.yarn.archive which will be more efficient using the distributed cache.

nchammas commented:
One use case I have for this feature is to make up for the fact that `spark.jars` and `spark.driver.extraClassPath` don't support relative paths. This is handy if you use Spark locally to run tests or to execute small-scale data tasks.

For example, say you have two projects on your local machine that each use PySpark.

```
project1/
  spark-conf/
    spark-defaults.conf
    jars/
      some.jar
  local-tool1.py
  requirements.txt
project2/
  spark-conf/
    spark-defaults.conf
    jars/
      another.jar
  local-tool2.py
  requirements.txt
```

You install/manage PySpark for each project via separate Python virtual environments. The projects need different versions of Spark, and depend on different jars and packages that Spark needs to be pointed to. You specify all of that via each project's spark-defaults.conf, and set `SPARK_CONF_DIR` depending on what project you're working on.

The problem is that you also want to point Spark at the appropriate jars, and spark-defaults.conf is not that helpful since you must specify absolute paths for `spark.jars` and `spark.driver.extraClassPath`. The absolute path to the jars will be different for each developer, since they can clone these projects to arbitrary directories, and since the absolute path will likely include their username in some way.

I believe `SPARK_JARS_DIR` solves this problem. We can remove any mention of `spark.jars` or `spark.driver.extraClassPath` from the defaults files, which are checked in to version control, and instead have people set `SPARK_JARS_DIR` depending on what project they are using.

I know this is not the use case that @marblejenka had in mind, but I wanted to share it since we have this problem at my current job. Perhaps this is more an argument for using Spark via Docker instead of virtual environments, but I wanted to share it for the record.

Commit:
modified to keep old lookup behavior in the case failed to find the dir specified by SPARK_JARS_DIR
renamed the argument name
modified to specify environment variable by constant in the testcase

SparkQA commented:
**[Test build #121743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121743/testReport)** for PR 28200 at commit [`fd9c804`](https://github.com/apache/spark/commit/fd9c80450f2be26565bd97c375f72f495c1becbd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26423/
Test PASSed.

marblejenka commented:
@tgravescs Certainly, we can do the same thing with SPARK_JARS_DIR by having multiple SPARK_HOME. One of the motivations for this PR is to have a single SPARK_CONF_DIR. Also, regarding spark.yarn.jars and spark.yarn.archive, it does not change the dependency of spark-shell process for example, so it is not sufficient for my usecase.
@nchammas Thank you for sharing your use case. However, I think that SPARK_DIST_CLASSPATH may work. SPARK_DIST_CLASSPATH allows you to add additional jars to your classpath. I'm not a user of pyspark, so there might be some misunderstanding.

Overall, SPARK_JARS_DIR is useful when you want to manage multiple sets of jars.

marblejenka commented:
@dongjoon-hyun I have fixed the points you have pointed out except for the additional dependency for testing. If adding new dependencies in this PR is unacceptable, I think opening another PR aimed at improving test coverage using environment variables is best so far.

SparkQA commented:
**[Test build #121743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121743/testReport)** for PR 28200 at commit [`fd9c804`](https://github.com/apache/spark/commit/fd9c80450f2be26565bd97c375f72f495c1becbd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121743/
Test FAILed.

tgravescs commented:
so normally I would think SPARK_CONF_DIR is separate from your spark home dir anyway. The production envs and hdp deployment I've seen were all separate from SPARK_HOME. Doesn't mean it has to be though.   I can see times it could be useful but other times it could be confusing.  Like someone updates spark to another version and updates the jars but doesn't get the changes for other components like pyspark/R.  But I guess that is a deployment issue.

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #121851 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121851/testReport)** for PR 28200 at commit [`fd9c804`](https://github.com/apache/spark/commit/fd9c80450f2be26565bd97c375f72f495c1becbd).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/26524/
Test PASSed.

dongjoon-hyun commented:
Thank you for updating, @marblejenka . 

According to the above discussion, it seems that the scope of this PR should be narrow downed. `spark-shell` is a good environment, but the majority of job submissions is `spark-submit` with `cluster` mode. And, @tgravescs pointed out, `spark.yarn.jars` and `spark.yarn.archive` is the popular way in YARN environment. Could you be more specific in the PR description about **the newly beneficial scope**? Is it beneficially only for `spark-shell`? Also, it would be great if you describe the alternatives like `spark.yarn.jars` and `spark.yarn.archive` together to be fair in the PR description.
> it does not change the dependency of spark-shell process for example, so it is not sufficient for my usecase.

SparkQA commented:
**[Test build #121851 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/121851/testReport)** for PR 28200 at commit [`fd9c804`](https://github.com/apache/spark/commit/fd9c80450f2be26565bd97c375f72f495c1becbd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/121851/
Test PASSed.

marblejenka commented:
I understand. Since I'm starting parental leave this week, I think the response will be this weekend.

marblejenka commented:
Due to family circumstances, I cannot spend time on this issue for a while. I think I can respond in the middle of May. I'm sorry.
